staging/src/k8s.io/apiserver/pkg/server/options/etcd.go

```diff
@@ -114,29 +108,46 @@ func (s *EtcdOptions) AddFlags(fs *pflag.FlagSet) {
 }
 
 func (s *EtcdOptions) ApplyTo(c *server.Config) error {
-	c.RESTOptionsGetter = &restOptionsFactory{options: s}
-
+	c.RESTOptionsGetter = RESTOptionsFactoryFunc(func(resource schema.GroupResource) (generic.RESTOptions, error) {
+		ret := generic.RESTOptions{
+			StorageConfig:           &s.StorageConfig,
+			Decorator:               generic.UndecoratedStorage,
+			EnableGarbageCollection: s.EnableGarbageCollection,
+			DeleteCollectionWorkers: s.DeleteCollectionWorkers,
+		}
+		if s.EnableWatchCache {
+			ret.Decorator = genericregistry.StorageWithCacher
+		}
+		return ret, nil
+	})
 	return nil
 }
 
-// restOptionsFactory is a default implementation of a RESTOptionsGetter
-// This will work well for most aggregated API servers.  The legacy kube server needs more customization
-type restOptionsFactory struct {
-	options *EtcdOptions
+func (s *EtcdOptions) ApplyWithStorageFactoryTo(factory serverstorage.StorageFactory, c *server.Config) error {
+	c.RESTOptionsGetter = RESTOptionsFactoryFunc(func(resource schema.GroupResource) (generic.RESTOptions, error) {
+		storageConfig, err := factory.NewConfig(resource)
+		if err != nil {
+			return generic.RESTOptions{}, fmt.Errorf("unable to find storage destination for %v, due to %v", resource, err.Error())
+		}
+
+		ret := generic.RESTOptions{
+			StorageConfig:           storageConfig,
+			Decorator:               generic.UndecoratedStorage,
+			DeleteCollectionWorkers: s.DeleteCollectionWorkers,
+			EnableGarbageCollection: s.EnableGarbageCollection,
+			ResourcePrefix:          factory.ResourcePrefix(resource),
+		}
+		if s.EnableWatchCache {
+			ret.Decorator = genericregistry.StorageWithCacher
+		}
+
+		return ret, nil
+	})
+	return nil
 }
 
-func (f *restOptionsFactory) GetRESTOptions(resource schema.GroupResource) (generic.RESTOptions, error) {
-	ret := generic.RESTOptions{
-		StorageConfig:           &f.options.StorageConfig,
-		Decorator:               registry.StorageWithCacher,
-		DeleteCollectionWorkers: f.options.DeleteCollectionWorkers,
-		EnableGarbageCollection: f.options.EnableGarbageCollection,
-		ResourcePrefix:          f.options.StorageConfig.Prefix + "/" + resource.Group + "/" + resource.Resource,
-	}
-
-	if !f.options.EnableWatchCache {
-		ret.Decorator = generic.UndecoratedStorage
-	}
+type RESTOptionsFactoryFunc func(resource schema.GroupResource) (generic.RESTOptions, error)
 
-	return ret, nil
+func (f RESTOptionsFactoryFunc) GetRESTOptions(resource schema.GroupResource) (generic.RESTOptions, error) {
```

Comment:
It's idiomatic, compare net/http handlers. But, I am fine with the struct if you hate it so much.